bucket/covergroup.py

```diff
@@ -34,21 +34,81 @@ def _chain_run(self, start: OpenLink[CovRun] | None = None) -> Link[CovRun]: ...
 class Covergroup(CoverBase):
     """This class groups coverpoints together, and adds them to the hierarchy"""
 
-    def __init__(self, name: str, description: str):
+    def __init__(self, name: str, description: str, subtree:str|list[str] = None):
         """
         Parameters:
             name: Name of covergroup
             description: Description of covergroup
-            cvg_tier: what coverage tier is being run [0..3]
+            subtree: Used to enable/disable coverpoints. 
+                     Should only be passed in for top covergroup
         """
```

Comment:
I think it would be better to have a separate method call rather than put this on the init, perhaps `prune()` or `filter()`?

Instead of passing round the list of subtrees you could pass round a match function that takes a coverpoint and returns true or false, that way the code down in `_apply_subtree` can be generic but you can have methods like `prune_by_fn()`, `prune_by_name()` or whatever which just create the match function and pass it through.

Reply:
I think an extra argument to filter which dictates what to do when a covergroup isn't matched would be good too - i.e whether to carry on checking the children and keep the covergroup alive if a child is alive. 
Perhaps `recurse_unmatched`?

Reply:
I imagined the subtree argument being passed in by the testbench when instantiating the top level covergroup. Could be split out, but makes for more code in the testbench to set everything up. The string(s) will either come from a command line argument, by the test case (or a combination of the two)
Passing in subtree during init would also allow us to re-implement how it's done later and prevent bucket goals being applied to inactive coverpoints. (Currently goals are applied before we determine if coverpoint is active)

If a covergroup isn't matched, it should always check its children for a match. I don't see why you would want it otherwise. Partial matches currently work, so don't have to provide a full path. Could just provide 'branches' and it would enable all covergroups/coverpoints that include 'branches'. If it only matched coverpoints, then all covergroups containing an active child would also get enabled. If it only matched a covergroup, then it would include all of its children as well. 

The internal sampling method checks to see if it is active before calling the user defined sample method. If a covergroup is active, it passes on the trace to its children. If it has determined that no children are active, it should be inactive itself, and therefore not call sample method of its children.

Reply:
Having it passed in through the init limits flexibility - if you want to add other forms of filtering you need to keep adding complexity to the arguments. Could always use method chaining (return self from the filter functions) if you want to keep it short e.g. 
```
myGroup = Covergroup("some_covergage", "cover stuff").subtree("xyz")
```

> If a covergroup isn't matched, it should always check its children for a match. I don't see why you would want it otherwise. 
I was thinking if in your coverage tree theres one covergroup with lots of slow coverage, you might want to disable everything inside that group - it's useful for the case where you want a blacklist not a whitelist.

Reply:
Depends on how much flexibility we want to offer. No reason we can't have an allow/deny list, but they will likely to be used differently. 

* Subtree is to be used for running particular regressions/tests quicker when debugging or targeting coverage. 
* Pruning is to be used to exclude particular coverpoints, such as when they have been fully hit in a prior regression, or a known list of larger/complicated coverpoints. 

Allow/Deny could be passed in as `dict{'allow': list[str], 'deny': list[str]}`, with deny taking priority over allow

Tiered coverage is supposed to be the way we exclude larger coverpoints for smaller regressions - just haven't implemented it yet!